### PR TITLE
Design Doc: The Client Library of Parameter Server

### PR DESCRIPTION
**Pull request by helinwang**
Maybe [here](https://github.com/helinwang/Paddle/blob/trainer_design/doc/design/cluster_train/pserver_client.md) is better for review.

### CODE REVIEW

File: doc/design/cluster_train/trainer.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+The Go interface is the basic abstraction of communications with the master server and parameter servers. We will add another layer on top (add retry logic, polish interface with C idiom) before exposing the library with a [C interface](#c-interface).
+
+```go
+// MasterClient is the client to the master server.
+type MasterClient struct {}
+
+// GetTask gets a new task by telling the master server the finished task.
+// Use nil as the finished task when getting the task for the first time.
+func (*MasterClient) GetTask(finished master.Task) (master.Task, error)
```
Comment (gongweibao) [inline doc/design/cluster_train/trainer.md]: `Finish a task`和`Get a new task`是否分成两个函数比较好？

File: doc/design/cluster_train/trainer.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	Int64
+	UInt64
+	Float32
+	Float64
+)
+
+// Parameter is a piece of data to sync with the parameter server.
+type Parameter struct {
+	Name        string
+	ElementType ElementType
```
Comment (gongweibao) [inline doc/design/cluster_train/trainer.md]: 变量名和变量的类型是否需要区分一下？

File: doc/design/cluster_train/trainer.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Design Doc: Trainer Communication Library
+
+For an overview of trainer's role, please refer to [distributed training design doc](README.md). In this design doc, we will discuss the trainer's communication library, which will manage communication with parameter servers and the [master server](master_server.md). The library will be implemented in [Go](https://golang.org/) and made available as a static or dynamic library with a C header file.
+
+## Go Interface
+
+The Go interface is the basic abstraction of communications with the master server and parameter servers. We will add another layer on top (add retry logic, polish interface with C idiom) before exposing the library with a [C interface](#c-interface).
+
+```go
```
Comment (typhoonzero) [inline doc/design/cluster_train/trainer.md]: Are these interfaces kind of duplicate to https://github.com/PaddlePaddle/Paddle/pull/1964? The communication interfaces are export by the server and will be called remotely by the client. So the server and the client need to use the same piece of interface definition code.

**Comment by typhoonzero** [comment]
Well, in general, there are two ways of implement a parameter server:

1. Pre-define all the possible optimization algorithms, and implement these algorithms inside parameter server. So the client just call the "algo-id" with gradients and some arguments.
1. Define basic OPs inside parameter server like "tensor add" or "tensor multiply" etc. The client then send gradients and a expression describing the optimization algorithm, eg. `error := SendGrad("momentom(param, grads, lita, ...)", grads)`

It's the first way in this document. Is the second choince possible?

File: doc/design/cluster_train/trainer.md (hunk 2 above)
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: 这倒不影响，因为ElementType作为变量名的调用方式是Parameter.ElementType，不会与全局scope的ElementType冲突。

File: doc/design/cluster_train/trainer.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// GetParam gets parameters from parameter servers.
+func (*PServerClient) GetParam(names []string) ([]Parameter, error)
+
+// Save indicates parameters to save the parameter to the given path.
+//
+// Path needs to be the path to a distributed file system which is visible
+// to all parameter servers.
+func (*PServerClient) Save(path string) error
+```
+Please see [master server design doc](master_server.md) for the definition of `master.Task`.
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: The file is in a PR. Will be a good link after the PRbeing merged.

File: doc/design/cluster_train/trainer.md (hunk 3 above)
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: The client library is not just a simple wrapper on top of RPC. For example `ParamInitChan` will use an etcd lock to figure out who will init the parameters.

Indeed, the API here highly depend on the parameter server RPC API. So this PR and #1964 depends on each other. This in the API in my mind, I will be discussing with @dzhwinter and everyone to form the same API, borrowing ideas from both PRs.

File: doc/design/cluster_train/trainer.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
+# Design Doc: Trainer Communication Library
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: Done.

File: doc/design/cluster_train/trainer.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+For an overview of trainer's role, please refer to [distributed training design doc](README.md). In this design doc, we will discuss the trainer's communication library, which will manage communication with parameter servers and the [master server](master_server.md). The library will be implemented in [Go](https://golang.org/) and made available as a static or dynamic library with a C header file.
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: Agree. I removed the master communication part.

File: doc/design/cluster_train/trainer.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+## Go Interface
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: I replaced the Go API with C API. Go Implementation can be discussed later.

File: doc/design/cluster_train/trainer.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+The Go interface is the basic abstraction of communications with the master server and parameter servers. We will add another layer on top (add retry logic, polish interface with C idiom) before exposing the library with a [C interface](#c-interface).
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: I replaced the Go API with C API.

File: doc/design/cluster_train/trainer.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	UInt64
+	Float32
+	Float64
+)
+
+// Parameter is a piece of data to sync with the parameter server.
+type Parameter struct {
+	Name        string
+	ElementType ElementType
+	Buffer      []byte
```
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: Changed to content (in C API, Go API has been removed.)

File: doc/design/cluster_train/trainer.md (hunk 1 above)
Comment (helinwang) [inline doc/design/cluster_train/trainer.md]: 我把Master部分删了，下次再发一个PR，会分成两个函数。

File: doc/design/cluster_train/pserver_client.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
+# Design Doc: The Client Library of Parameter Server
+
+For an overview of trainer's role, please refer to [distributed training design doc](README.md). In this design doc, we will discuss the parameter server's client library, which will manage communication with parameter servers. The library will be implemented in [Go](https://golang.org/) and made available as a static or dynamic library with a C header file.
+
+## C Interface
+
+```c
+#define PADDLE_ELEMENT_TYPE_INT32   0
```
Comment (gongweibao) [inline doc/design/cluster_train/pserver_client.md]: define => enum?

File: doc/design/cluster_train/pserver_client.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+#define PADDLE_ELEMENT_TYPE_UINT32  1
+#define PADDLE_ELEMENT_TYPE_INT64   2
+#define PADDLE_ELEMENT_TYPE_UINT64  3
+#define PADDLE_ELEMENT_TYPE_FLOAT32 4
+#define PADDLE_ELEMENT_TYPE_FLOAT64 5
+
```
Comment (gongweibao) [inline doc/design/cluster_train/pserver_client.md]: Do we need to specify a namespace?

File: doc/design/cluster_train/pserver_client.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+typedef struct {
```
Comment (gongweibao) [inline doc/design/cluster_train/pserver_client.md]: Do we need to specify a namespace?

File: doc/design/cluster_train/pserver_client.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * @brief paddle_send_grads sends gradients to parameter servers for
+ * updating parameters.
+ *
+ * @param grads the array of gradients to send.
+ * @param total the total number of gradient inside the gradient array.
+ * @param learning_rate the learning rate for the gradients.
+ * @return 0 if successful, otherwise -1.
+ */
+int paddle_send_grads(paddle_pserver_client* client, const paddle_gradient* grads, int total, double learning_rate);
```
Comment (dzhwinter) [inline doc/design/cluster_train/pserver_client.md]: maybe use the same interface with "set_parameter" better?  otherwise, the set_parameters only used once.

File: doc/design/cluster_train/pserver_client.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
```
Comment (dzhwinter) [inline doc/design/cluster_train/pserver_client.md]: in fact, I  want to know where the Updater/optimizer should be. 
The trainer need to calculate gradients before send grads, and the pserver also need optimizer to store the training state

**Comment by reyoung** [comment]
Is how to update sparse parameter considered in this PR?
Is how the regularization works considered in this PR?

File: doc/design/cluster_train/pserver_client.md (hunk 12 above)
Comment (helinwang) [inline doc/design/cluster_train/pserver_client.md]: c里面貌似没有namespace，一般用前缀来防止link时候的冲突，比如我们用了`paddle_`。

File: doc/design/cluster_train/pserver_client.md (hunk 14 above)
Comment (helinwang) [inline doc/design/cluster_train/pserver_client.md]: Actually here it's only sending the learning rate, not the update method. The update method is sent [here](https://github.com/PaddlePaddle/Paddle/pull/2075/files/c583447e900af61ac69fa6b0a41e9ab10cf3e90a#diff-6b7fb9815373bb8e7487da918861823aR62) using Proto Buffer.

File: doc/design/cluster_train/pserver_client.md (hunk 10 above)
Comment (helinwang) [inline doc/design/cluster_train/pserver_client.md]: Thanks! Done.

File: doc/design/cluster_train/pserver_client.md (hunk 13 above)
Comment (helinwang) [inline doc/design/cluster_train/pserver_client.md]: Removed the set_parameter interface, only keeping paddle_send_grads. I think set_parameter complicates stuff.
As far as I know, set_parameter is only used for setting the learning rate (or optimizer related stuff). I think such use case does not justify the complication added to the system:

1. What should we do when saving parameters, do we saves these parameters as well?
1. Make the interface harder to understand.

If the user wants to use a different learning rate the next time he starts a training job, he can just change the python code.

Maybe there are other use cases for set_parameter that I am unaware of?

**Comment by helinwang** [comment]
@reyoung 谢谢提醒！如何支持Sparse更新以及正则化的支持我都加到design doc里了，请看更新版本。
另外还总结了些，写到了：https://github.com/PaddlePaddle/Paddle/issues/2106#issuecomment-300944231

File: doc/design/cluster_train/pserver_client.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
+# Design Doc: The Client Library of Parameter Server
+
+For an overview of trainer's role, please refer to [distributed training design doc](README.md). In this design doc, we will discuss the parameter server's client library, which will manage communication with parameter servers. The library will be implemented in [Go](https://golang.org/) and made available as a static or dynamic library with a C header file.
+
+## Parameter Partition
+
+Each parameter will be partitioned into parameter chunks to make the parameters evenly distributed on parameter servers. The partition is done automatically by the client library. The *sparse parameter* require a little different treatment:
```
Comment (dzhwinter) [inline doc/design/cluster_train/pserver_client.md]: Hmm, we should unify the Parameter Partition name with ParameterServer, I just wrote `ParameterBlock` as the Paddle already have.

Comment (helinwang) [inline doc/design/cluster_train/pserver_client.md]: Done. Renamed parameter chunks to parameter blocks.
